pkg/skaffold/kubernetes/portforward/resource_forwarder_test.go

```diff
@@ -252,3 +259,101 @@ func TestUserDefinedResources(t *testing.T) {
 		}
 	})
 }
+
+func mockClient(m kubernetes.Interface) func() (kubernetes.Interface, error) {
+	return func() (kubernetes.Interface, error) {
+		return m, nil
+	}
+
+}
+
+func TestRetrieveServices(t *testing.T) {
+	tests := []struct {
+		description string
+		namespaces  []string
+		services    []*v1.Service
+		expected    []*latest.PortForwardResource
+	}{
+		{
+			description: "multiple services in multiple namespaces",
+			namespaces:  []string{"test", "test1"},
+			services: []*v1.Service{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "svc1",
+						Namespace: "test",
+						Labels: map[string]string{
+							deploy.RunIDLabel: "9876-6789",
+						},
+					},
+					Spec: v1.ServiceSpec{Ports: []v1.ServicePort{{Port: 8080}}},
+				}, {
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "svc2",
+						Namespace: "test1",
+						Labels: map[string]string{
+							deploy.RunIDLabel: "9876-6789",
+						},
+					},
+					Spec: v1.ServiceSpec{Ports: []v1.ServicePort{{Port: 8081}}},
+				},
+			},
+			expected: []*latest.PortForwardResource{{
+				Type:      constants.Service,
+				Name:      "svc1",
+				Namespace: "test",
+				Port:      8080,
+				LocalPort: 8080,
+			}, {
+				Type:      constants.Service,
+				Name:      "svc2",
+				Namespace: "test1",
+				Port:      8081,
+				LocalPort: 8081,
+			}},
+		}, {
+			description: "no services in given namespace",
+			namespaces:  []string{"randon"},
+			services: []*v1.Service{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "svc1",
+						Namespace: "test",
+						Labels: map[string]string{
+							deploy.RunIDLabel: "9876-6789",
+						},
+					},
+					Spec: v1.ServiceSpec{Ports: []v1.ServicePort{{Port: 8080}}},
+				},
+			},
+		}, {
+			description: "services present but does not expose any port",
```

Comment:
is this possible for a service?

Reply:
you wont normally do it. But yes, you can not have any ports defined in the Spec and the service is created.

```
$ kubectl apply -f microservices/leeroy-app/kubernetes/
service/leeroy-app created
deployment.apps/leeroy-app created

$ git diff microservices/leeroy-app/kubernetes/
diff --git a/examples/microservices/leeroy-app/kubernetes/deployment.yaml b/examples/microservices/leeroy-app/kubernetes/deployment.yaml
index 56ef7152..46feab6d 100644
--- a/examples/microservices/leeroy-app/kubernetes/deployment.yaml
+++ b/examples/microservices/leeroy-app/kubernetes/deployment.yaml
@@ -6,9 +6,6 @@ metadata:
     app: leeroy-app
 spec:
   clusterIP: None
-  ports:
-    - port: 50051
-      name: leeroy-app
   selector:
     app: leeroy-app
 ---
tejaldesai@@examples (move_resolve_namespace)$ 
```